### PR TITLE
NETOBSERV-669: overview export PNG

### PR DESCRIPTION
**Pull request by Amoghrd**
Export overview page and each panel in overview page
Update topology export function

### CODE REVIEW

**Comment by Amoghrd** [comment]
@jpinsonneau Could you please review this when you are free

File: web/src/components/netflow-overview/panel-kebab.tsx [hunk 1]
```diff
@@ -58,6 +60,23 @@ export const PanelKebab: React.FC<PanelKebabProps> = ({ id, options, setOptions
...
+        .then(dataUrl => {
+          const link = document.createElement('a');
+          link.download = 'overview.png';
+          link.href = dataUrl;
+          link.click();
+        })
+        .catch(err => {
+          console.log(err);
+        });
+  }, [setOptions, id]);
```
Comment (jpinsonneau) [inline web/src/components/netflow-overview/panel-kebab.tsx:78]:
```suggestion
  }, [id, isDark]);
```

We need to update the hook dependency here according to its content :wink:

File: web/src/components/netflow-overview/panel-kebab.tsx [hunk 2]
```diff
@@ -58,6 +60,23 @@ export const PanelKebab: React.FC<PanelKebabProps> = ({ id, options, setOptions
...
 
+  const onOverviewExport = React.useCallback(() => {
+    const overview_flex = document.getElementById(id);
+
+    setShowOptions(false);
+    if (overview_flex)
+      toPng(overview_flex, { cacheBust: true, backgroundColor: isDark ? '#0f1214' : '#f0f0f0' })
+        .then(dataUrl => {
+          const link = document.createElement('a');
+          link.download = 'overview.png';
```
Comment (jpinsonneau) [inline web/src/components/netflow-overview/panel-kebab.tsx:71]:
```suggestion
          link.download = `overview_${id}.png`;
```

What do you think about reusing the panel id in the exported file name ?

File: web/src/components/netflow-overview/panel-kebab.tsx [hunk 3]
```diff
@@ -58,6 +60,23 @@ export const PanelKebab: React.FC<PanelKebabProps> = ({ id, options, setOptions
...
+          link.href = dataUrl;
+          link.click();
+        })
+        .catch(err => {
+          console.log(err);
```
Comment (jpinsonneau) [inline web/src/components/netflow-overview/panel-kebab.tsx:76]:
```suggestion
          console.error(err);
```

to show an error in the browser console

File: web/src/components/netflow-overview/panel-kebab.tsx [hunk 4]
```diff
@@ -58,6 +60,23 @@ export const PanelKebab: React.FC<PanelKebabProps> = ({ id, options, setOptions
     [setOptions, options]
   );
 
+  const onOverviewExport = React.useCallback(() => {
+    const overview_flex = document.getElementById(id);
```
Comment (jpinsonneau) [inline web/src/components/netflow-overview/panel-kebab.tsx:64]: nipticking here but I see there is a margin on right + bottom in the exported screen

overview


```suggestion
    const overview_flex = document.getElementById(id)?.children[0] as HTMLElement | undefined;
```
taking the first child would remove it

overview without margin

**Review by jpinsonneau — COMMENTED** [review]
Small changes but LGTM :partying_face: thanks @Amoghrd !

Do you want to centralize the `toPng` calls in a single place like `utils/export.ts` ? 
It would avoid repeating the options, the link creation and the catch. That can be a followup if you prefer.

**Comment by Amoghrd** [comment]
@jpinsonneau When implementing the `utils/export.ts` change, I am getting 
`<e> [webpack-dev-server] [HPM] Error occurred while proxying request localhost:9001/api/loki/flows?filters=&limit=50&recordType=flowLog&reporter=destination&timeRange=300&rateInterval=30s&step=15s&type=count to http://localhost:9002/ [ECONNREFUSED] (https://nodejs.org/api/errors.html#errors_common_system_errors)
<e> [webpack-dev-server] [HPM] Error occurred while proxying request localhost:9001/api/frontend-config to http://localhost:9002/ [ECONNREFUSED] (https://nodejs.org/api/errors.html#errors_common_system_errors)
<e> [webpack-dev-server] [HPM] Error occurred while proxying request localhost:9001/api/prometheus/api/v1/rules?type=alert to http://localhost:9002/ [ECONNREFUSED] (https://nodejs.org/api/errors.html#errors_common_system_errors)
<e> [webpack-dev-server] [HPM] Error occurred while proxying request localhost:9001/api/alertmanager/api/v2/silences to http://localhost:9002/ [ECONNREFUSED] (https://nodejs.org/api/errors.html#errors_common_system_errors)
<e> [webpack-dev-server] [HPM] Error occurred while proxying request localhost:9001/api/loki/topology?filters=&limit=5&recordType=flowLog&reporter=destination&timeRange=300&rateInterval=30s&step=15s&type=bytes&scope=namespace to http://localhost:9002/ [ECONNREFUSED] (https://nodejs.org/api/errors.html#errors_common_system_errors)
<e> [webpack-dev-server] [HPM] Error occurred while proxying request localhost:9001/api/loki/topology?filters=&limit=5&recordType=flowLog&reporter=destination&timeRange=300&rateInterval=30s&step=15s&type=bytes&scope=app to http://localhost:9002/ [ECONNREFUSED] (https://nodejs.org/api/errors.html#errors_common_system_errors)
<e> [webpack-dev-server] [HPM] Error occurred while proxying request localhost:9001/api/loki/ready to http://localhost:9002/ [ECONNREFUSED] (https://nodejs.org/api/errors.html#errors_common_system_errors)
`
when I do make-standalone-mock. Could you PTAL and let me know if there's something wrong in the utils file

**Comment by jpinsonneau** [comment]
> @jpinsonneau When implementing the `utils/export.ts` change, I am getting `<e> [webpack-dev-server] [HPM] Error occurred while proxying request

@Amoghrd sorry about that. I didn't checked the impact of https://github.com/netobserv/network-observability-console-plugin/pull/315/files on standalone mode that actually conflicts.
I [forced the metrics port to 9003](https://github.com/netobserv/network-observability-console-plugin/pull/322/commits/e9a9d6394eb95e27bfb04f495c831d3db0d9c50f) for this specific usage.

Also took the opportunity to [refactor a bit the export function ](https://github.com/netobserv/network-observability-console-plugin/pull/322/commits/a0f7e1104afee22dc16fd39b98dd850774ad5ff3)
- renamed to `exportToPng` 
- check the HTMLElement inside it
- log error is provided element is undefined

**Pull request merged**